### ISSUE TITLE
[2002.07971] Gradient Boosting Neural Networks: GrowNet

### ISSUE THREAD

**Issue by riron1206** [riron1206/read_paper#30]
## 概要
https://arxiv.org/abs/2002.07971
1,2層程度の浅いMLPを弱モデルとしてブースティング

![nn](https://raw.githubusercontent.com/sbadirli/GrowNet/master/Model.png)

## 新規性・差分
ブースティングとMLPの組み合わせ

## 手法
浅いMLPを1epoch学習→最終層の出力を特徴量列に追加→浅いMLPを1epoch学習→… を繰り返す。論文では40回程度繰り返してた

## 結果
分類、回帰、ランキングのタスクにおいてGBDTより精度高い

## コメント
- 公式のgithub(pytorch): https://github.com/sbadirli/GrowNet
- マルチラベル分類での使用例: https://www.kaggle.com/tmhrkt/grownet-gradient-boosting-neural-networks

<br>

---------------------------------------------------------------------

<br>

## 論文翻訳メモ

勾配ブーストにおける重要な考え方は、単純で低次のモデルを弱い学習者として取り上げ、それを基本的なビルディングブロックとして使用して、1次または2次の勾配統計量を使用した逐次ブーストによって強力で高次のモデルを構築することです。
本論文では、弱い学習者として浅いニューラルネットワーク（例えば、1つまたは2つの隠れ層を持つ）を使用する。
ブースティングの各ステップでは、元の入力特徴量を現在の反復の最後の層からの出力で増強する（図1参照）。
この増強された特徴セットは、現在の残差を用いたブースティング機構を介して次の弱い学習者を訓練するための入力として供給される。
モデルの最終的な出力は、これらの連続して訓練されたモデルのスコアの重み付けされた組み合わせとなります。

GrowNetは出力を予測するためにK個の加算関数を使用します。
各関数ｆｋは、線形層を出力層とする独立した浅いニューラルネットワークを表す

与えられたサンプル x に対して、モデルは GrowNet で fk の重み付けされた和として予測を計算

モデルの複雑さにペナルティを与えるために正則化項を追加することもできますが、この作業では単純化のために省略します。
我々が最適化しようとしている目的は関数に対してであり、パラメータに対してではないので、従来の最適化手法はここでは機能しません

GBDT [11]と同様に、モデルは加法的に学習されます。
サンプルxiに対するt-1段階でのGrowNetの出力を y? (t-1) i = Pt-1 k=0 αkfk(xi)とします。
我々は貪欲に次の弱い学習者ft(x)を求め、ステージtでの損失を最小化します。(2) 
また、計算の複雑さを緩和するために損失関数lのテイラー展開を採用した。
2 次最適化手法は 1 次最適化手法よりも優れていることが証明されており、収束までに必要なステップ数も少ないため、ニュートンラプソンステップを用いてモデルを学習しています。
その結果、MLタスクに関わらず、個々のモデルパラメータはGrowNetの出力の2次勾配への回帰を実行することで最適化されます。
↑2次勾配への回帰の問題に変えてるので回帰のlossも使う

弱い学習者ftの目的関数は、L (t) = Xn i=0 hi(?yi - αtft(xi))2 (3) のように単純化することができ、
ここでy?i = -gi/hi、giとhiはxiにおける目的関数lの1次と2次の勾配であり、
y? (t-1) i. (補足資料よりアルゴリズム1のパート1の疑似コードを参照)

t番目の弱い学習者のパラメータのみがブーストステップtで更新されることを意味し、
前のt - 1の弱い学習者のパラメータはすべて変更されない。
↑これではモデルが局所的な最小値に陥る可能性があり、固定の昇圧率αkはこの問題を悪化させるので、
修正ステップでは、前回のt -1の弱い学習者を固定するのではなく、
バックプロパゲーションによって前回のt -1の弱い学習者のパラメータを更新できるようにした。
さらに、ブースト率αkをモデルのパラメータに組み込み、修正ステップで自動的に更新する。
これにより、性能が向上するだけでなく、デリケートなパラメータを調整する必要がなくなりました。

2 次の弱い学習者ftは、{xi , y?i}の2次勾配統計量を用いた最小二乗回帰によって適合される。
修正ステップでは、2値クロスエントロピー損失を用いてモデル全体を再訓練することにより、追加されたすべての予測関数のパラメータが更新されます。

実験セットアップ。
モデルに追加された予測関数はすべて2つの隠れ層を持つ多層パーセプトロンである。
隠れ層の単位数は一般的に入力特徴量の約半分かそれ以上に設定します。
隠れ層の数を増やすと，モデルがオーバーフィットし始めると性能が低下する．
実験では3つのタスクすべてに40個の加法関数（MLPを40用意してるということ）を採用し、
検証結果からテスト時間内の弱い学習者の数を選択した。
ブースティング率は初期に1に設定し、修正ステップ中に自動的に調整した。
各予測関数を1エポックだけ訓練し、全体のモデルも修正ステップ中に1エポックだけアダムオプティマイザーを用いた確率的勾配降下法で訓練した。
ランキングタスクでは、エポック数を2に増やした。隠れ層には2次元バッチ正規化を採用した。
LightGBMやCatBoost、AdaNetでも同様の結果が得られるため、XGBoostとの比較を行った。
3つの手法のチューニングとモデルの詳細は補足資料を参照されたい。

分類 XGBoostとの公平な比較のために、XGBoostの論文[6]で使用されているHiggs bozonデータセットを用いて、我々のモデルをテストした。
分類結果を表3に示します。GrowNetは、すべてのデータを用いてXGBoostを明らかに凌駕しています。
データの10%をサブサンプリングして弱い学習者を学習させることでも、性能が向上しています。
30 個の弱学習器（16 個の隠れ層を持つ多層パーセプトロン）を使用し、検証結果に応じてテスト時に使用する弱学習器の数を選択しています。
3つの実験では、この数はすべて30であった。

モデルのすべての構成要素の中で，補正ステップはおそらく最も重要なものである．
このステップでは，モデルに追加されたすべての弱い学習者のパラメータが，最後の層の特徴を含まない元の入力でモデル全体を訓練することによって更新される．
このステップで使用される損失関数はタスク固有のものである。
この手順により、モデルは負の勾配に合わせるのではなく、手元のタスクに合わせてパラメータを修正することができます。
C/Sはまた、弱い学習者間の潜在的な相関を緩和する。
さらに、このステップでは、ブースト率αtを組み込み、チューニングを必要とせずに自動的に調整されます。
表4の最後の2列は、GrowNetの分類結果とランク付け学習結果を、補正ステップを使用しない場合と、5段階ごとに補正ステップを使用した場合の結果をそれぞれ示しています。
前者では性能が著しく低下しており、予測関数を追加した後はほとんど情報を学習しませんでした。
図2のフラットな学習損失もこの現象を裏付けています。
補正ステップを5段階ごとに実行すると、性能ははるかに向上しましたが、GrowNetの結果ほどではありませんでした。

動的なブースト率。補正ステップの中で、ステージtでのブースト率αtを動的に更新することができます。
この対策をとることで、パラメータの調整の手間を省くことができ、すべてのタスクでマイルドな性能向上を実現しました。
また、ブーストレートを固定した場合と比較して、学習損失の収束性が向上しました（図2参照）。
我々の設定では，α0 = 1 から，補正ステップを実行するたびにブースト率が自動的に更新されます（図 3）．0.01, 0.1, 1}のセットで粗く調整した最高の一定ブーストレート（αt = 0.1）の結果を表4の5列目に報告する。
